builder/csharp_builder.rb

```diff
@@ -53,13 +53,18 @@ def list_disposable_files
   end
 
   def create_project(mand_files, disp_files)
-    tmpl = File.read(@project_template)
-    files_xml = (mand_files + disp_files).map { |x| "    <Compile Include=\"#{x}\" />" }.join("\n")
-    project = tmpl.gsub(/%%%FILES%%%/, files_xml)
+    tmpl = File.read(@project_file)
+    files_xml = (disp_files).map { |x| "    <Compile Include=\"#{x}\" />" }.join("\n")
+    project = tmpl.gsub(/    <Compile Include="..\\..\\..\\compiled\\csharp\\\*\*\\\*.cs" \/>/, files_xml)
     File.write(@project_file, project)
```

Comment:
What you propose here is generally a very bad thing to do: you programmatically *edit* files checked into the repository. Please keep them separated: either we use checked-in .csproj file without any changes (changing the environment around it, for example), or we clearly show that it's a template and we're generating actual .csproj every time we build it with this builder.

Reply:
I agree that it is bad to modify committed files. But I cannot come up with anything better.

> either we use checked-in .csproj file without any changes (changing the environment around it, for example)

Can you propose how to do it without the template? 

I do not like the template, because using it forces a developer to manually change it and keep track of it not to commit to the repo. Also, a developer cannot open the solution and build it out-of-the-box after cloning it. He needs to figure it out.

I think that it is best to allow a developer to clone and run it out of the box the same way as he is used to on .net platform and does necessary modifications on CI.

Reply:
> I do not like the template, because using it forces a developer to manually change it and keep track of it not to commit to the repo.

Why would developer need to manually change a template? The very point of using a template is separation of "template" (which is normally never changed) and "project" (which is changed frequently, being generated in the build process, depending on how well stuff compiles).

> Also, a developer cannot open the solution and build it out-of-the-box after cloning it. He needs to figure it out.

Developer cannot open that solution in any case. At the bare minimum, workflow requires (1) building compiler, (2) building ksy -> csharp files, (3) either cleaning up project manually to exclude things which fail to compile, or invoking partial builder to generate that project for you.

> I think that it is best to allow a developer to clone and run it out of the box the same way as he is used to on .net platform

We support more than a dozen of different languages/platforms. While it's an interesting goal to make KS feel more "natural" for certain target audiences, I don't think we'll realistically be able to satisfy everyone. If we'll aggressively move towards the flow that .NET developers are used to (e.g. integrating compiler / format building into csproj), that might hurt the flow which others are used to.

Reply:
> Why would developer need to manually change a template? The very point of using a template is separation of "template" (which is normally never changed) and "project" (which is changed frequently, being generated in the build process, depending on how well stuff compiles).

>Developer cannot open that solution in any case. At the bare minimum, workflow requires (1) building compiler, (2) building ksy -> csharp files, (3) either cleaning up project manually to exclude things which fail to compile, or invoking partial builder to generate that project for you.

Well, my point was to limit dependencies that a developer needs to install on his machine before he can contribute. The bare minimum required Java, Scala, Ruby, etc. So it is not just a "clone" operation for building and contribution.

> We support more than a dozen of different languages/platforms. While it's an interesting goal to make KS feel more "natural" for certain target audiences, I don't think we'll realistically be able to satisfy everyone. If we'll aggressively move towards the flow that .NET developers are used to (e.g. integrating compiler / format building into csproj), that might hurt the flow which others are used to.

I did not meant to bend KS for .NET developers, but for any developer in any language that KS supports, so a developer would not be required to install dependencies beyond a setup that is common for his language. It might be a troublesome in corporates when you need approvals for SW installation so it might discourage people. 

But I see your point so let me to change it back to the template.